### ISSUE TITLE
[filter_attack] type= matches unexpected attack type

### ISSUE THREAD

**Issue by soliton-** [wesnoth/wesnoth#8585]
When attacking with a weapon that has multiple types via `[damage_type] alternative_type=` an attack event with `[filter_attack] type=` matches for the original and alternative damage type.

IMO the expectation is that only the actually used type is matched. So if the weapon for example is pierce and the alternate type from the weapon special is arcane then when attacking undead the filter should match `type=arcane` (the type actually used in the attack) not `type=pierce`.

An example of this issue currently happens in NR where you can use holy water to get weapons to have arcane as alternate type.

If a unit with a pierce weapon that used a holy potion attacks Malifor this case is triggered:
https://github.com/wesnoth/wesnoth/blob/f9927cb6035c3a76d5a0e6aa0fc23e0eae210476/data/campaigns/Northern_Rebirth/scenarios/05a_01_The_Pursuit.cfg#L2755
even though the attack does use arcane. (You can't see arcane triggering as well because the arcane handling was wrongly removed, see #8584)

**Comment by newfrenchy83** [comment]
when [damage_type]alternative_type used then the choice of type is valable for numrical calculation of damage, but filter don't acced to resistance of opponent for could doing thi choice, for filtering the type is both pierce and arcane(it is yhat what holy water should doing, add 'arcane' type to a weapon who has already own type)

**Comment by newfrenchy83** [comment]
ifyou want what only arcane matched when using double type use
`[filter_attack]
type=pierce
    [not]
        type=arcane
    [/not]
[/filter_attack]
`

**Comment by soliton-** [comment]
> when [damage_type]alternative_type used then the choice of type is valable for numrical calculation of damage, but filter don't acced to resistance of opponent for could doing thi choice, for filtering the type is both pierce and arcane(it is yhat what holy water should doing, add 'arcane' type to a weapon who has already own type)

It is so frustrating how I need to read every message from you at least 5 times to have any chance to begin to guess what you're trying to say.

Anyway, it sounds like you're saying that attack filters need to be improved if they have no idea what the attack did. I agree. That is why I opened this issue.



> ifyou want what only arcane matched when using double type use
> 
> > [filter_attack]
> 
> > type=pierce
> 
> > [not]
> 
> > type=arcane

I don't want that.

**Comment by CelticMinstrel** [comment]
I think this is expected. From what I understand, `alternative_type` means that the weapon now has both the new type and the old, so it should match both in filtering.

How exactly can the filter know which type actually ends up being used in a given context?

**Comment by soliton-** [comment]
Ends up? The player just picked a weapon and attacked...

**Comment by CelticMinstrel** [comment]
How does the filter know about that?

**Comment by newfrenchy83** [comment]
> How does the filter know about that?

It don't know, if attack has two type and one type same what filter, it is matches same if this is type what doing lesser damage on opponent

**Comment by soliton-** [comment]
In case it may not be clear, attack events run after the attack is confirmed by the player. It is already decided by the game what type the attack uses as was shown in the attack dialog. The information is available.

Thus it is not expected or makes any sense to match multiple types. Only the actually used type should match.

**Comment by newfrenchy83** [comment]
> In case it may not be clear, attack events run after the attack is confirmed by the player. It is already decided by the game what type the attack uses as was shown in the attack dialog. The information is available.
> 
> Thus it is not expected or makes any sense to match multiple types. Only the actually used type should match.

But i can't make calculation of resistanceof opponent in attack_type.cpp matches_simple_filter(), in filter_attack, it is not the damage inflicted who are pertinent but if one of damage type matches with filter, because alternative_type give a double type to weapon, the original and the secondary, an sword with Holy water will inflict both blade and arcane but it is the type who make more damage who will be used in damage calculation, but for events both types can be checked

**Comment by soliton-** [comment]
> But i can't make calculation of resistanceof opponent in attack_type.cpp matches_simple_filter(), in filter_attack, it is not the damage inflicted who are pertinent but if one of damage type matches with filter, because alternative_type give a double type to weapon, the original and the secondary, an sword with Holy water will inflict both blade and arcane but it is the type who make more damage who will be used in damage calculation, but for events both types can be checked

Why is this a single sentence...

Regardless of whatever implementation issues there may or may not be. One more time: the information is there. It makes no sense to match multiple types for an event based on an already picked damage type. Only the one picked type should match.

**Comment by newfrenchy83** [comment]
It's not "the enemy resists cutting so I use the arcane instead", it's "I use a sharpened sword coated with holy water; but it's the latter which inflicts the the most significant damage". We then have a simultaneous use of the two types in terms of event and if one of them is filtered we have a correspondence but in the calculation of the damage inflicted on the enemy, we only retain the most effective type for the sake of simplicity for the player.

So the fact that you say that only the chosen type should be filtered makes no sense to me. In addition it could be an advantage such as triggering an event that requires a dual type attack (an opponent who only suffers damage from a physical/magical attack for example)

**Comment by soliton-** [comment]
> It's not "the enemy resists cutting so I use the arcane instead", it's "I use a sharpened sword coated with holy water; but it's the latter which inflicts the the most significant damage". We then have a simultaneous use of the two types in terms of event and if one of them is filtered we have a correspondence but in the calculation of the damage inflicted on the enemy, we only retain the most effective type for the sake of simplicity for the player.

So the attack behaves one way but that is somehow not the real thing so attack event logic works differently? No idea how that is expected or useful.

Do you understand that the behavior in NR right now simply does not make sense? How do you propose to fix that with the current filter logic?

> So the fact that you say that only the chosen type should be filtered makes no sense to me. In addition it could be an advantage such as triggering an event that requires a dual type attack (an opponent who only suffers damage from a physical/magical attack for example)

I say that the attack event should work according to how the attack behaves.

How does this dual type attack event work? Does it have anything to do with how `[filter_attack]` currently works?

**Comment by ProditorMagnus** [comment]
> "I use a sharpened sword coated with holy water; but it's the latter which inflicts the the most significant damage"
If it is not allowed to say that 40% of [attack] unmodified damage is type a and 60% of unmodified damage type b, then it is not useful to complicate filtering. If calculation uses only 1 type then so should filter.

**Comment by newfrenchy83** [comment]
> > "I use a sharpened sword coated with holy water; but it's the latter which inflicts the the most significant damage" If it is not allowed to say that 40% of [attack] unmodified damage is type a and 60% of unmodified damage type b, then it is not useful to complicate filtering. If calculation uses only 1 type then so should filter.

fixed in https://github.com/wesnoth/wesnoth/pull/8936 PR

**Comment by doofus-01** [comment]
>>    So the fact that you say that only the chosen type should be filtered makes no sense to me. In addition it could be an advantage such as triggering an event that requires a dual type attack (an opponent who only suffers damage from a physical/magical attack for example)

>I say that the attack event should work according to how the attack behaves.

I agree with soliton- and ProditorMagnus, but also see where newfrenchy83 is coming from.  It seems like this multiple damage types per attack is an anti-KISS mistake.

**Comment by Jonathan-Kelly** [comment]
Closing this as fixed by #8936
